### PR TITLE
cmd/bigslice: extract build and run into a package to allow for reuse.

### PR DESCRIPTION
**Pull request by cosnicolaou**
This is a protoype for one means of more easily integrating bigslice into other user specific environments.

This PR extracts the bigslice commands 'build' and 'run' functions into a separate package to allow them to be reused in other user environments. The intent being to allow user sites to customize their setup by providing their own bigslice command, with a site specific setup-ec2 command whilst still reusing the rest of the bigslice commands functionality.

### CODE REVIEW

File: cmd/bigslice/build.go [hunk 1]
```diff
@@ -75,181 +29,16 @@ func buildCmd(args []string) {
 	flags.Usage = func() { buildCmdUsage(flags) }
 	must.Nil(flags.Parse(args))
 
+	if len(flags.Args()) == 0 {
+		log.Fatal("no arguments")
+	}
+
+	// TODO(cosnicolaou): this seems wrong to me, surely it should just be
+	//   flags.Args()?
```
Comment (mariusae) [inline cmd/bigslice/build.go]: yeah that looks like a bug

**Comment by cosnicolaou** [comment]
> How do you imagine a custom `bigslice` tool would inject site-specific configuration? Through build variables? Inlining a profile?
> 
> I'm sure you discovered the hidden/undocumented `BigsliceMain` feature in the bigslice tool. That was part of the thinking behind that: by having a custom entry point the tool can inject site-specific configuration into the built binary; but I decided to go the route of profiles instead.

What I've done is provide my own implementation of the bigslice command where build/run use the package as created here, but instead I have custom setup commands that create profiles appropriate for our environments. The usage is thus:

```
go run neeva.co/cmd/bigslice setup this-ec2-instance
```
or
```
go run neeva.co/cmd/bigslice setup this-laptop
```

this generates a neeva-ec2 profile get used by:

```
go run neeva.co/cmd/bigslice build -output binary ....
```
etc.

Make sense?

**Comment by mariusae** [comment]
I see. Though in this case, the only thing you're really replacing is how the profile is written? Seems like you could just distribute the appropriate profile directly?

**Comment by cosnicolaou** [comment]
On Mon, Oct 21, 2019 at 11:17 AM marius a. eriksen <notifications@github.com>
wrote:

> I see. Though in this case, the only thing you're really replacing is how
> the profile is written? Seems like you could just distribute the
> appropriate profile directly?
>
yes, but sadly the profile is different for every user and depends on their
environment also. For example, I'll set it up so that logs access is only
possible from ec2 machines in a restricted vpc/subnet vs general dev which
is available more broadly etc.

Cheers, Cos.

**Comment by mariusae** [comment]
Ok, that makes sense. Though even in this case it could be provided as an orthogonal command? 

```
$ bigslice-neeva setup ... 
```

and then just using the `bigslice` tool for everything else. (Indeed, if you're running on Linux, the use of `bigslice` should be entirely optional.)

**Comment by cosnicolaou** [comment]
well sure, but that's two things to explain to my users etc. And with my
approach I can use my command line tools, flags etc and just make
everything consistent.

Cheers, Cos.

**Comment by mariusae** [comment]
That's fair. I'm ok with this change, as it doesn't really add any complexity, but I'd really like to figure out a way to make this much simpler...

File: cmd/bigslice/build.go (hunk 1 above)
Comment (mariusae) [reply cmd/bigslice/build.go]: want to fix this while we're at it?

**Comment by cosnicolaou** [comment]
do you have any convenient means of testing the bigslice command with the [1:] removed?

**Comment by mariusae** [comment]
Fixing the build now; then will merge.

**Pull request merged**